core/src/main/java/com/linecorp/armeria/common/stream/AsyncMapStreamMessage.java

```diff
@@ -0,0 +1,241 @@
+            upstream = subscription;
+            downstream.onSubscribe(this);
+        }
+
+        @Override
+        public void onNext(T item) {
+            requireNonNull(item, "item");
+
+            if (canceled) {
+                StreamMessageUtil.closeOrAbort(item);
+                return;
+            }
+
+            try {
+                final CompletableFuture<U> future = function.apply(item);
+                requireNonNull(future, "function.apply() returned null");
+
+                pendingRequests++;
+                future.handle((res, cause) -> {
+                    if (executor.inEventLoop()) {
+                        publishDownstream(res, cause);
+                    } else {
+                        executor.execute(() -> publishDownstream(res, cause));
+                    }
+                    return null;
+                });
+            } catch (Throwable ex) {
+                StreamMessageUtil.closeOrAbort(item, ex);
+                upstream.cancel();
+                onError(ex);
+            }
+        }
+
+        private void publishDownstream(@Nullable U item, @Nullable Throwable cause) {
+            if (canceled) {
+                if (item != null) {
+                    StreamMessageUtil.closeOrAbort(item);
+                }
+
+                return;
+            }
+
+            try {
+                if (cause != null) {
+                    upstream.cancel();
+                    onError(cause);
+                } else {
+                    requireNonNull(item, "function.apply()'s future completed with null");
+                    downstream.onNext(item);
+
+                    if (--pendingRequests == 0 && isCompleting) {
+                        downstream.onComplete();
+                    } else if (requestedByDownstream.get() > 0) {
+                        if (requestedByDownstream.get() != Long.MAX_VALUE) {
+                            requestedByDownstream.decrementAndGet();
+                        }
+                        upstream.request(1);
+                    }
+                }
+            } catch (Throwable ex) {
+                StreamMessageUtil.closeOrAbort(item, ex);
+                upstream.cancel();
+                onError(ex);
+            }
+        }
+
+        @Override
+        public void onError(Throwable cause) {
+            requireNonNull(cause, "cause");
+            if (canceled) {
+                return;
+            }
+            canceled = true;
+
+            downstream.onError(cause);
+        }
+
+        @Override
+        public void onComplete() {
+            if (canceled) {
+                return;
+            }
+
+            if (pendingRequests > 0) {
+                isCompleting = true;
+            } else {
+                downstream.onComplete();
+            }
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0) {
+                onError(new IllegalArgumentException(
```

Comment:
How about aborting `source` instead of calling `onError` so that `source` does not produce the elements anymore?

Reply:
My understanding is that it has to call onError according to rule 3.9 of the reactive streams spec
> Subscription.request(long n) MUST signal onError with a java.lang.IllegalArgumentException if the argument is <= 0.

Reply:
Ah, I didn't realize calling abort would pass the exception down. I'll fix it

Reply:
Oh, but since this is in the subscriber we don't have access to the source. I think it's better to call onError and cancel in that case